### PR TITLE
Add Schematron Validation

### PR DESCRIPTION
**Pull request by Firionus**
Hey, 

I recently reworked the XML Schema to provide support for DIN 15800:2020-07 (GDTF 1.1). Due to the restricted expressiveness of XSD 1.0, there were some validation gaps left in the Schema that I wanted to fill. This can be done with an assertion-based schema language like [Schematron](https://schematron.com/). I added a Schematron file called `gdtf.sch`. It can be used to validate instance files with different tools, I used the [schXslt](https://github.com/schxslt/schxslt) Java CLI for development. 

I downloaded all GDTF 1.1 files from GDTF Share on January 27 2021 20:00. All of those XML files pass the Schematron validation proposed here. 

Included validations:
- Restricted Occurence for some children of PhysicalDescriptions/Properties
- Validate Feature Reference in Attribute Definition
- Validate ModeMaster Reference in ChannelFunction
- Unique DMXChannel name
- Unique LogicalChannel

# Remaining Questions to the Maintainer(s)

- How is the default name of the ChannelFunction constructed? Like this:
```
Dimmer1
Pan2
Pan3
```
or like this:
```
Dimmer1
Pan1
Pan2
```
- What does "All logical channels that are children of the same DMX channel are exclusive. " (DIN 15800:2020-07, page 33) mean?
- How is it ensured that in a given DMXMode, no two ChannelFunctions controlling the same Attribute of the same Geometry are active at the same time, therefore providing potentially two different values for the same Attribute?

### CODE REVIEW

**Comment by Firionus** [comment]
> How is the default name of the ChannelFunction constructed?

So I did the smart thing and opened up the builder: 
image
It seems like the default Name of a ChannelFunction is the Attribute, followed by a space and a number. This number is the 1-based index of the ChannelFunction inside this LogicalChannel. I'll try to incorporate it into the Schematron. 
Should I open an issue that this needs to be clarified in the standard?

**Comment by petrvanekrobe** [comment]
> > How is the default name of the ChannelFunction constructed?
> 
> So I did the smart thing and opened up the builder:
> It seems like the default Name of a ChannelFunction is the Attribute, followed by a space and a number. This number is the 1-based index of the ChannelFunction inside this LogicalChannel. I'll try to incorporate it into the Schematron.
> Should I open an issue that this needs to be clarified in the standard?

Thank you.

The name should just be unique, there is no need for any defined way to construct it. User can also change the name as needed, to reflect actual behavior...

**Comment by Firionus** [comment]
> > > How is the default name of the ChannelFunction constructed?
> > 
> > 
> > So I did the smart thing and opened up the builder:
> > It seems like the default Name of a ChannelFunction is the Attribute, followed by a space and a number. This number is the 1-based index of the ChannelFunction inside this LogicalChannel. I'll try to incorporate it into the Schematron.
> > Should I open an issue that this needs to be clarified in the standard?
> 
> Thank you.
> 
> The name should just be unique, there is no need for any defined way to construct it. User can also change the name as needed, to reflect actual behavior...

Thanks for the comment, I'll have to disagree though. The standard defines a Default value for this attribute. It's not very clear from just the standard what this default is, but it means the attribute "Name" for a ChannelFunction is optional. If the attribute is missing, the default value needs to be assumed. Even if the Name attribute is missing, the ChannelFunction can be referenced as a ModeMaster via its default Name. Therefore, the default name is not arbitrary. 
In short: If the Name attribute is omitted and the ChannelFunction is referenced as a ModeMaster, it is important that the default Name be clearly defined.

**Comment by moritzstaffel** [comment]
The Specs defines the channel names like this:

>Name
>Unique name; Default value: Name of attribute and number of channel function.

So @Firionus is right with this. What about the following definition

>Name
>Unique name **within the Channel Function**; Default value: Name of attribute and number of channel function **inside the Channel Function**

**Comment by Firionus** [comment]
What about:
>Name
>Unique name within the Logical Channel; Default value: Name of Attribute and 1-based index of Channel Function inside the Logical Channel, separated by a space; Example: "Dimmer 1".

- Fixed mix-up of Channel Function and Logical Channel
- Made clear both identifiers are to be separated by a space
- Clarified that 1-based indexing is used (Computer Scientists have a way of always assuming zero-based indexing...)
- Added Example which should help understanding
- Standardized capitalization of words like "Channel Function", but this should be adjusted to be consistent throughout the standard

**Comment by petrvanekrobe** [comment]
>Name
>Unique name within the Channel Function; Default value: Name of attribute and number of channel function inside the Channel Function

Are you suggesting that both names can coexist at the same time? Thus:

`<ChannelFunction Attribute="Gobo2" DMXFrom="0/1" Default="0/1" Name="RotatingGobo index">`

Can be reference both with:

`...ModeMaster="head_Gobo2.Gobo2.RotatingGobo index"...`

and

`...ModeMaster="head_Gobo2.Gobo2.Gobo2 1"...`

At the same time?

Also what about this: as an author, can create two Channel Functions:

`<ChannelFunction Attribute="Gobo2" Name="">`
`<ChannelFunction Attribute="Gobo2" Name="Gobo2 1">`

These are unique Names (thus valid GDTF) but now cannot be uniquely referenced as per your example above where the name is generated on the fly _during usage_, because functions both have the same name. I just tested this in the builder and it generates this xml for me:

```
<ChannelFunction Attribute="Gobo2" Name="">
<ChannelFunction Attribute="Gobo2"  Name="Gobo2 1">
```

So where is this following Channel Function's Master pointing at?

`<ChannelFunction Attribute="Gobo2Pos" ModeMaster="head_Gobo2.Gobo2.Gobo2 1">`

**Comment by Firionus** [comment]
> Are you suggesting that both names can coexist at the same time?

No, since Default values are only used if the attribute is absent. If the attribute is set, the set value is used. That is how XML Schema does ist and how I would understand the DIN. 

> These are unique Names (thus valid GDTF) but now cannot be uniquely referenced as per your example above where the name is generated on the fly during usage, because functions both have the same name.

I already put this on my ToDo list. Schematron can enforce that the user does not set a Name attribute which clashes with an existing Default value. 
Of course, this would have to be added to the builder and whatever other proprietary software you already have for validation. 
The simples solution would be to require the Name attribute in a future version (breaking change!). I think the Builder always includes the Name attribute anyway, so there would be no change required to the Builder. 

As for your example, Petr:

```
<ChannelFunction Attribute="Gobo2" Name="">
<ChannelFunction Attribute="Gobo2"  Name="Gobo2 1">
```
> So where is this following Channel Function's Master pointing at?

```
<ChannelFunction Attribute="Gobo2Pos" ModeMaster="head_Gobo2.Gobo2.Gobo2 1">
```

This ModeMaster is pointing at the second one with the name "Gobo2 1". The first one has the name "" (empty string). The default value is not applied here because the attribute is present. The default value is only used if the attribute is absent.

**Comment by petrvanekrobe** [comment]
Hi @Firionus ,

thank you for all the work on this.

Are there still any other open questions you have, before this can PR can be finalized and merged?

Also, as you have seen some issues when scrolling through the spec, please feel free to report any inconsistency in the issue tracker.

Thank you :)

Petr

**Comment by Firionus** [comment]
One questions from the initial post is still open:

> What does "All logical channels that are children of the same DMX channel are exclusive. " (DIN 15800:2020-07, page 33) mean?



I also think I found an answer to another question from the initial post:

> How is it ensured that in a given DMXMode, no two ChannelFunctions controlling the same Attribute of the same Geometry are active at the same time, therefore providing potentially two different values for the same Attribute?

It isn't.  
GDTF-utilizing Software needs to resolve such conflicts heuristically. Or throw up their hands and tell the user they don't know what to do with this GDTF file.  

The underlying issue is that GDTF does not properly define the semantics of LogicalChannel and ChannelFunction. The format allows *very* complex descriptions of ChannelFunctions without defining what each case means for a console or visualizer. 

I guess I'll open up an issue with some examples where the semantics are unclear? 

Should I also open an issue about the suggestion in https://github.com/mvrdevelopment/spec/pull/64#issuecomment-779761060?

**Comment by petrvanekrobe** [comment]
> I guess I'll open up an issue with some examples where the semantics are unclear?

Yes please.

> Should I also open an issue about the suggestion in [#64 (comment)](https://github.com/mvrdevelopment/spec/pull/64#issuecomment-779761060)?

That would be great,

thank you!

Petr

**Comment by Firionus** [comment]
Okay, I guess we can merge then. 

The XPath code isn't really a thing of beauty. It's the first time I've written XPath, so the result is... functional but may need some polishing down the road.

**Comment by Firionus** [comment]
After 8 months, I think it is time to follow up on this PR. 

I would like to close it without merging. I don't think anyone really wants to learn Schematron (me included), or any other XML validation language for that matter. The validation that can't be done with XSD 1.0 is probably better done in the parser which people can understand rather than some obscure XML thing. 

I'd suggest putting a comment into the XSD file, indicating the parts that aren't validated in the XSD but have to be checked at the parser level. Does that sound good?

**Comment by petrvanekrobe** [comment]
Hi @Firionus , this makes sense to me,

Cheers
P.

**Comment by Firionus** [comment]
I opened #104 to close this PR.

Cheers back @petrvanekrobe :wave: ,
Johannes
